File: eea.cpp
#include "eea.h"
// extended euclidian algorithm to find the modular inverse of "a" modulo modulus
// we store the modular inverse always as > 0
largeInt eea(const largeInt& a, const largeInt& modulus){
    largeInt r = modulus;
    largeInt newR = a % modulus;

    largeInt t = 0ULL;
    largeInt newT = 1ULL;

    while (newR != largeInt(0ULL)) {
        largeInt quotient = r / newR;

        largeInt nextR = r % newR;

        largeInt product = (quotient * newT) % modulus;
        largeInt nextT = modSubtract(t, product, modulus);

        r = newR;
        newR = nextR;

        t = newT;
        newT = nextT;
   
    }

    if (r != largeInt(1ULL)) {
        throw std::domain_error("Modular inverse does not exist");
    }

    return t;
}

File: pythonAlgos/eea.py
#EXTENDED EUCLIDEAN ALGORITHM IN PYTHON

def gcdExtended(a, b):
    if a == 0:
        return b, 0, 1
    gcd, x1, y1 = gcdExtended(b % a, a)
    x = y1 - (b // a) * x1
    y = x1
    return gcd, x, y


#print("GCD is", g)
#print("x =", x, ", y =", y)
#x is always the modular inverse of a modulo b







    

